8.2.1 suspended ports ja

niku opened the pull request.

niku committed:
8.2.1 suspended ports ja

niku commented on 108-cpu-ja.tex:
```diff
@@ -80,15 +80,18 @@ \section{システムモニター}
 \label{subsec:port-system-monitors}
 
-An interesting part of system monitors that didn't fit anywhere but may have to do with scheduling is regarding ports. When a process sends too many message to a port and the port's internal queue gets full, the Erlang schedulers will forcibly de-schedule the sender until space is freed. This may end up surprising a few users who didn't expect that implicit back-pressure from the VM.
+%An interesting part of system monitors that didn't fit anywhere but may have to do with scheduling is regarding ports. When a process sends too many message to a port and the port's internal queue gets full, the Erlang schedulers will forcibly de-schedule the sender until space is freed. This may end up surprising a few users who didn't expect that implicit back-pressure from the VM.
```
> An interesting part of system monitors that didn't fit anywhere but may have to do with scheduling is regarding ports. 

ここまでの節のどこにもうまく書けるところがなかったけど，でもスケジューリングに関することとして〜
という意味であると解釈しましたが，どうでしょうか．(何に対しての fit なのかよくわからなかった)

hykw replied:
LGTM
that の前に動詞がないので、that は関係代名詞つまり X that didn't fit = X didn't fit ということですね。

hykw commented on 108-cpu-ja.tex:
```diff
@@ -80,15 +80,18 @@ \section{システムモニター}
 
-An interesting part of system monitors that didn't fit anywhere but may have to do with scheduling is regarding ports. When a process sends too many message to a port and the port's internal queue gets full, the Erlang schedulers will forcibly de-schedule the sender until space is freed. This may end up surprising a few users who didn't expect that implicit back-pressure from the VM.
+%An interesting part of system monitors that didn't fit anywhere but may have to do with scheduling is regarding ports. When a process sends too many message to a port and the port's internal queue gets full, the Erlang schedulers will forcibly de-schedule the sender until space is freed. This may end up surprising a few users who didn't expect that implicit back-pressure from the VM.
+これまでのどの文にもあてはまらなかったけれどスケジューリングに関係のある、システムモニターのおもしろいところはポートに関するところです。あるプロセスが沢山のメッセージをポートに送りポートの内部キューが一杯になったとき、スペースに空きがでるまでErlangのスケジューラーは強制的に送り側のスケジュールを解除します。これはVMから暗黙のバックプレッシャーを予想していなかった少数のユーザーを驚かせることになるかもしれません。
```
「少数のユーザ」にちょっと違和感を感じました。

ここの a few は「多数や普通の数よりも少ない」という多→少の視点というよりは、「0ではなく少しはいる」という0→小の視点で、かつ「そんな人いないだろー」という著者の皮肉っぽい感情も含んでるので、「このことは、VMによる暗黙のバックプレッシャーを予想もしていなかったユーザを驚かすかもしれません」ぐらいさらっと訳したほうがいいかも、です。

niku committed:
8.2.1 一時停止しているポートの翻訳を改善

ma2gedev commented on 108-cpu-ja.tex:
```diff
@@ -80,15 +80,18 @@ \section{システムモニター}
+こういったイベントは \term{busy\_port} アトムをシステムモニターへ送ることでモニターできます。特にクラスタ化されたノードで、ノード間の通信がビジーポートによって処理されていた場合、リモートノード上のプロセスと通信しているローカルプロセスがスケジュール解除されるため、それを発見するのに \term{busy\_dist\_port} アトムが使われます。
 
+%If you find out you're having problems with these, try replacing your sending functions where in critical paths with \function{erlang:port\_command(Port, Data, [nosuspend])} for ports, and \function{erlang:send(Pid, Msg, [nosuspend])} for messages to distributed processes. They will then tell you when the message could not be sent and you would therefore have been descheduled.
+もしそういった問題を抱えていることを発見した場合、クリティカルパスにある送信の関数をポートに送る場合は \function{erlang:port\_command(Port, Data, [nosuspend])}、分散したプロセスの場合は \function{erlang:send(Pid, Msg, [nosuspend])} へと置き換えてみてください。それらはメッセージを送れないためスケジュール解除されるときに知らせてくれます。
```
> それらはメッセージを送れないためスケジュール解除されるときに知らせてくれます。

「それらはメッセージを送れない時とスケジュール解除されている時に知らせてくれます。」という感じではないでしょうか。前の文章でクラスタ化の場合は de-scheduled とあったので繋がっているのかなと考えました。

niku replied:
なるほど．ほとんどの文に自信がないので意見をいただけてありがたいです ❤️ 

ma2gedev さんの解釈は when A and B という構造に注目して
メッセージが送れなかった and スケジュールが解除された ときに知らせてくれる
というものですかね．

私は文をこういう形に分けて訳しました．

`(They will then tell you) (when ((the message could not be sent) and (you would therefore have been descheduled))).`

その中でも therefore に注目していました．

therefore https://eow.alc.co.jp/search?q=therefore
その［前に述べたことの］結果（として）

であるそうなので

> you would therefore have been descheduled
前に述べたことの結果としてスケジュール解除される

前に述べたこととは

> the message could not be sent
メッセージを送ることができなくなった

だと考えました．

繋げると
メッセージが送れなくなったためにスケジュール解除されたときに知らせてくれる
というものです．

また最初の文が could 次の文が would でずらしているのは因果関係がそこにはあるのかなと思いましたがこれは自信ないです……

もう一回考えてみます．

ma2gedev replied:
お手間かけてしまいすみません。私の解釈も分解していただいてありがとうございます 🙇 
お考えを聞いて確かに niku さんの言われている通り therefore で繋がっていると考える方が自然だと思いました。

hykw replied:
`They will then tell you when the message could not be sent and you would therefore have been descheduled.`

could が過去形なのは仮定法だからですね。would have been はいわゆる大過去と言われるやつです。will → would なのは時制の一致か仮定法ですがどっちも有り得るしどっちでもいい。

メッセージが送れないようなことがあれば（仮定法）、その結果としてデスケジュールされるようなことがあったら（could not be sent よりも、さらに過去の話）、tell you するよ、みたいな意味かと。

shino replied:
@hykw さんのコメントどおりなのですが、英語の話ではなくて Erlang の API 側からのコメントです。

nosuspend オプションは呼び出し側をブロックさせないことが目的だと記憶しています。
http://erlang.org/documentation/doc-8.0/erts-8.0/doc/html/erlang.html#send-3

    If the sender would have to be suspended to do the send, nosuspend is returned instead.

    [適当な訳] 呼び出し側が送信しようとするとサスペンドさせられてしまうのであれば、 nosuspend が返る

雰囲気としては EWOULDBLOCK 的な感じで。
そのため「deschedule されたら知らせてくれる」だと結局ブロックしちゃうように読めないことがないので、「deschedule されるような状態になったばあいは(ブロックせずに)(なんかしらエラーを)知らせてくれる」という意図が分かる訳になると良いな、と思います。

niku replied:
皆様ありがとうございます。

それらはメッセージを送れないためスケジュール解除されるときに知らせてくれます。 ->
メッセージを送れないことでスケジュール解除される場合すぐに知らせてくれます。

のように変えてみました。
「すぐに」にブロックしないという意図をこめたつもりです。
訳注で明記する案も考えましたが、これでいかがでしょうか。

niku committed:
訳改善

> They will then tell you when the message could not be sent and you would therefore have been descheduled.

それらはメッセージを送れないためスケジュール解除されるときに知らせてくれます。 ->
メッセージを送れないことでスケジュール解除される場合すぐに知らせてくれます。

shino commented:
LGTM 👍
